<a href="https://colab.research.google.com/github/naye971012/2023_sanhak_ML_study/blob/main/dealing_keyword_in_csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
%cd /content/drive/MyDrive/ColabNotebooks/산학_테스트

/content/drive/MyDrive/ColabNotebooks/산학_테스트


#import module

In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 48.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.8/236.8 kB 22.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 83.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 67.5 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
import torch
import torch.nn as nn

In [ ]:
from transformers import AutoTokenizer, AutoModel

#erasing 중복

In [ ]:
xml_sample = pd.read_csv('xml_sample_20230519.csv')

xml_sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3074 entries, 0 to 3073
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   template_idx           3074 non-null   int64 
 1   page_no                3074 non-null   int64 
 2   thumbnail_url          3074 non-null   object
 3   sheet_url              3074 non-null   object
 4   title                  3074 non-null   object
 5   primary_colors         3074 non-null   object
 6   primary_color_weights  3074 non-null   object
 7   is_video_type          3074 non-null   bool  
 8   language               3074 non-null   object
 9   keyword                3074 non-null   object
dtypes: bool(1), int64(2), object(7)
memory usage: 219.3+ KB


In [ ]:
keyword = xml_sample['keyword']
keyword

0       밤|노래|뮤직|새벽|음악|인디|재생|발라드|새벽밤|인기곡|감성힙합|새벽노래|플레이리...
1       교회|기도|예배|찬양|기도회|기독교|섬네일|썸네일|유튜버|유튜브|찬양예배|유튜브섬네...
2             깔금|세카|심플|영상|예배|종교|비대면|온라인|세션카드|온라인예배|송구영신예배
3       과일|과즙|기록|네일|미용|방송|셀프|손톱|심플|여행|일기|일상|커플|키위|토크|페...
4       네온|해안|섬네일|썸네일|유튜브|드라이브|사이버펑크|유튜브섬네일|유튜브썸네일|플레이리스트
                              ...                        
3069    개인|고객|미소|부업|소개|여가|채널|취미|퍼플|귀여운|마케터|마케팅|보라색|알파벳...
3070    개인|고객|미소|부업|소개|여가|채널|취미|퍼플|귀여운|마케터|마케팅|보라색|알파벳...
3071    개인|고객|미소|부업|소개|여가|채널|취미|퍼플|귀여운|마케터|마케팅|보라색|알파벳...
3072    개인|고객|미소|부업|소개|여가|채널|취미|퍼플|귀여운|마케터|마케팅|보라색|알파벳...
3073                개인|소비|여가|절약|취미|홍보|소셜 미디어|마케팅|절약방|커뮤니티
Name: keyword, Length: 3074, dtype: object

In [ ]:
keyword[0]

'밤|노래|뮤직|새벽|음악|인디|재생|발라드|새벽밤|인기곡|감성힙합|새벽노래|플레이리스트|새벽에듣기좋은노래'

In [ ]:
keyword_sum = []

for i in range(len(keyword)):
  keyword_sum+=keyword[i].split('|')

keyword_sum = list(set(keyword_sum))

print(len(keyword_sum) )

print(keyword_sum[:20])

4176
['소박한', '회사', '건강식', '교과', '은빛', '외출금지', '교육 협회', '영어', '웨딩', '오스카상', '화이트데이', '기계', '실적', '상담센터', 'f', '각진', '할인특가', '개월', '어워드', 'DIY']


#calculate similirity

In [ ]:
tokenizer = AutoTokenizer.from_pretrained("monologg/kobert")
model = AutoModel.from_pretrained("monologg/kobert")

In [ ]:
count_unk = 0

for i in keyword_sum:
  input_word = i
  tokens = tokenizer.tokenize(input_word)
  if(tokens[0]=='[UNK]'):
    count_unk+=1
print(count_unk)

3704


In [ ]:
"""
#tokens = ["[CLS]"] + tokens + ["[SEP]"]

input_ids = tokenizer.convert_tokens_to_ids(tokens)

print(input_ids)

input_tensor = torch.tensor([input_ids])

outputs = model(input_tensor)

word_embedding = outputs[0][:,0,:].detach().numpy()

#word_embedding.shape

"""

결론. 4176개의 키워드를 koBert를 이용해 embedding하고, KNN을 통해 유사한 단어를 제거하려고 하였지만 UNK, 즉 모델이 학습하지 않은 단어가 3700개로 매우 많아 해당 방법이 불가능함.